### DISCUSSION TITLE
converting between AD<double> and AD<complex double>

### DISCUSSION THREAD

**Discussion by nicoleruckstuhl** [coin-or/CppAD#134] [Q&A]
Hi, I'm trying to convert my code to use CppAD. However my code makes heavy use of complex specific functions (such as conj(), norm(), arg() ) which don't seem to be incorporated in CppAD. Is this correct? How could they be added? 
I believe I would be able to do everything I need with the following 3 functions:
AD<double> = real(AD<complex double>)
AD<double> = imag(AD<complex double>)
AD<comple double> = cd(AD<double>, AD<double>)
Am I correct in understanding that these functions to convert between AD<double> and AD<complex double> don't exist in CppAD? Would it be possible to add them? If yes could you point me to how to do that, or to an example on how to convert from one AD base type to another? 
Thank you

**Comment by bradbell** [comment]
The functions real, imag and conj are not complex differentiable. They are however real differentiable. If you used `complex < AD<double>` > instead of `AD< complex<double> >` it should work. The problem is that std::complex is only guarenteed to work for the base types float, double, and long double, so you may have to implement your own version of 
```template <class Type> class complex```

**Comment by nicoleruckstuhl** [comment]
Thank you, I'll try to continue that way.
